kv/kvserver: TestNodeLivenessRetryAmbiguousResultError failed

kv/kvserver.TestNodeLivenessRetryAmbiguousResultError [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15744489?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/15744489?buildTab=artifacts#/) on release-23.2.7-rc @ [9545977a4c6e69d452ac03fab6604ee0d2f4cd8f](https://github.com/cockroachdb/cockroach/commits/9545977a4c6e69d452ac03fab6604ee0d2f4cd8f):


```
=== RUN   TestNodeLivenessRetryAmbiguousResultError
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestNodeLivenessRetryAmbiguousResultError520138237
    test_log_scope.go:81: use -show-logs to present logs inline
    node_liveness_test.go:1005: expected injected error count of at least 2; got 1
    node_liveness_test.go:1007: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestNodeLivenessRetryAmbiguousResultError520138237
--- FAIL: TestNodeLivenessRetryAmbiguousResultError (6.25s)
```

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestNodeLivenessRetryAmbiguousResultError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-39678

--- a/pkg/kv/kvserver/node_liveness_test.go
+++ b/pkg/kv/kvserver/node_liveness_test.go
@@ -994,19 +994,21 @@ func TestNodeLivenessRetryAmbiguousResultError(t *testing.T) {
 	testutils.SucceedsSoon(t, func() error {
 		return verifyLivenessServer(s, 1)
 	})
-	nl := s.NodeLiveness().(*liveness.NodeLiveness)
-
-	l, ok := nl.Self()
-	assert.True(t, ok)
 
 	// And again on manual heartbeat.
+	// NOTE: we make sure to set pause the heartbeat loop before we grab the
+	// liveness record to ensure that we don't race with the heartbeat loop in
+	// a way that allows our manual heartbeat to short-circuit.
+	nl := s.NodeLiveness().(*liveness.NodeLiveness)
+	defer nl.PauseHeartbeatLoopForTest()
+
 	injectError.Store(true)
-	if err := nl.Heartbeat(context.Background(), l); err != nil {
-		t.Fatal(err)
-	}
-	if count := injectedErrorCount.Load(); count < 2 {
-		t.Errorf("expected injected error count of at least 2; got %d", count)
-	}
+	l, ok := nl.Self()
+	assert.True(t, ok)
+	require.NoError(t, nl.Heartbeat(context.Background(), l))
+
+	// Verify that the error was injected exactly twice.
+	require.Equal(t, int32(2), injectedErrorCount.Load())
 }
 
 // This tests the create code path for node liveness, for that we need to create

--- a/pkg/kv/kvserver/node_liveness_test.go
+++ b/pkg/kv/kvserver/node_liveness_test.go
@@ -964,17 +964,16 @@ func TestNodeLivenessRetryAmbiguousResultError(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	var injectError atomic.Value
-	var injectedErrorCount int32
+	var injectError atomic.Bool
+	var injectedErrorCount atomic.Int32
 
 	injectError.Store(true)
 	testingEvalFilter := func(args kvserverbase.FilterArgs) *kvpb.Error {
 		if _, ok := args.Req.(*kvpb.ConditionalPutRequest); !ok {
 			return nil
 		}
-		if val := injectError.Load(); val != nil && val.(bool) {
-			atomic.AddInt32(&injectedErrorCount, 1)
-			injectError.Store(false)
+		if injectError.Swap(false) {
+			injectedErrorCount.Add(1)
 			return kvpb.NewError(kvpb.NewAmbiguousResultErrorf("test"))
 		}
 		return nil
@@ -1005,7 +1004,7 @@ func TestNodeLivenessRetryAmbiguousResultError(t *testing.T) {
 	if err := nl.Heartbeat(context.Background(), l); err != nil {
 		t.Fatal(err)
 	}
-	if count := atomic.LoadInt32(&injectedErrorCount); count < 2 {
+	if count := injectedErrorCount.Load(); count < 2 {
 		t.Errorf("expected injected error count of at least 2; got %d", count)
 	}
 }

